sdk/nodejs/runtime/rpc.ts

```diff
@@ -494,3 +543,20 @@ export function deserializeProperty(prop: any): any {
         return obj;
     }
 }
+
+type ProxyConstructor = {
+    new(name: string, args: any, opts: { urn: string }): Resource;
+};
+
+const proxyConstructors = new Map<string, ProxyConstructor>();
+/**
+ * registerProxyConstructor registers a constructor to be used as a proxy for any URNs matching the
+ * given type that are deserialized by the current instance of the Pulumi JavaScript SDK.
+ */
+export function registerProxyConstructor(type: string, constructor: ProxyConstructor) {
```

Comment:
How will we handle versioning?

It seems like it might be better to perform registration on a per-package basis, then call into the package and let it deal with constructing the type. This would require fewer and simpler registrations--just a package + version for any given package piece of a type (e.g. `registerPackage("aws", "1.2.3", someConstructCallback)`).

Reply:
> How will we handle versioning?

What versioning scenario do you imagine coming into play here?  

Is it SxS versions of the NPM modules?  If so, I assume that will work fine as each of the modules will maintain it's own registered constructors it will use to dehydrate.  

Or is it about a skew between the client and server libraries (server registers with v1, client attempts to read and deserialize that into v2)?  In that case, it's possible we should try to identify this case, but it's not clear we can do any true version negotiation - the package author has to ensure they deliver matching (or compatible) versions of the libraries on both sides.

> It seems like it might be better to perform registration on a per-package basis

Curious why?  That seems to complicate codegen, and I don't see an obviously beneficial reason for it?   Also - what would `someConstructCallback` itself do - does it return a big list of registrations?  Ultimately I kind of like this being decoupled from the notion of packages unless there is a need to couple it.